### PR TITLE
Revamp Security context to be more specific and expose it to task templates

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
Introduce SecurityContext message at TaskTemplate, LaunchPlan and Execution levels to allow customization of security attributes; the identity to run as, the secrets to get access to, the tokens to issue...

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Tracking Issue
https://github.com/flyteorg/flyte/issues/800

### CODE REVIEW

File: protos/flyteidl/core/types.proto [hunk 1]
```diff
@@ -98,3 +98,102 @@ message Error {
     // Error message thrown.
     string message = 2;
 }
+
```
Comment (kumare3) [inline protos/flyteidl/core/types.proto:101]: can we move them to a different file, like common.proto or security.proto?

File: protos/flyteidl/core/types.proto [hunk 2]
```diff
@@ -98,3 +98,102 @@ message Error {
...
+// Secret encapsulates information about the secret a task needs to proceed. An environment variable
+// _FLYTE_SECRET_ENV_PREFIX will be passed to indicate the prefix of the environment variables that will be present if
+// secrets are passed through environment variables.
+// _FLYTE_SECRET_PATH_PREFIX will be passed to indicate the prefix of the path where secrets will be mounted if secrets
+// are passed through file mounts.
+message Secret {
+    enum Type {
```
Comment (kumare3) [inline protos/flyteidl/core/types.proto:108]: why is this necessary?

File: protos/flyteidl/core/types.proto [hunk 3]
```diff
@@ -98,3 +98,102 @@ message Error {
     // Error message thrown.
     string message = 2;
 }
+
+// Secret encapsulates information about the secret a task needs to proceed. An environment variable
+// _FLYTE_SECRET_ENV_PREFIX will be passed to indicate the prefix of the environment variables that will be present if
+// secrets are passed through environment variables.
+// _FLYTE_SECRET_PATH_PREFIX will be passed to indicate the prefix of the path where secrets will be mounted if secrets
```
Comment (kumare3) [inline protos/flyteidl/core/types.proto:105]:
```suggestion
// FLYTE_SECRETS_DEFAULT_DIR will be passed to indicate the prefix of the path where secrets will be mounted if secrets
```
this follows the convention in flytekit

File: protos/flyteidl/core/types.proto [hunk 4]
```diff
@@ -98,3 +98,102 @@ message Error {
     // Error message thrown.
     string message = 2;
 }
+
+// Secret encapsulates information about the secret a task needs to proceed. An environment variable
+// _FLYTE_SECRET_ENV_PREFIX will be passed to indicate the prefix of the environment variables that will be present if
```
Comment (kumare3) [inline protos/flyteidl/core/types.proto:103]:
```suggestion
// FLYTE_SECRETS_ENV_PREFIX will be passed to indicate the prefix of the environment variables that will be present if
```

File: protos/flyteidl/core/types.proto (hunk 2 above)
Comment (kumare3) [reply protos/flyteidl/core/types.proto:108]: I mean is there a use to know the type of secret? can we actually handle this correctly

File: protos/flyteidl/core/security.proto [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+        // Asymmetric indicates a public-private key where the reference is to the private portion of the key.
+        Asymmetric = 1;
+    }
+
+    // The name of the secret to mount. This has to match an existing secret in the system. It's up to the implementation
+    // of the secret management system to require case sensitivity.
+    string name = 1;
+
+    // Type indicates the type of key requested. Asymmetric keys can not be passed through environment variables. If the
```
Comment (kumare3) [inline protos/flyteidl/core/security.proto:25]: this is not true for Confidant for example - right? So instead do we want to say that envvar injection possible or no?

File: protos/flyteidl/core/security.proto [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    // k8s_service_account references a kubernetes service account to impersonate.
+    string k8s_service_account = 2;
+
+    // oauth2_client references an oauth2 client. Backend plugins can use this information to impersonate the client when
+    // making external calls.
+    OAuth2Client oauth2_client = 3;
+}
+
+// OAuth2TokenRequest encapsulates information needed to request an OAuth2 token.
+// FLYTE_TOKENS_ENV_PREFIX will be passed to indicate the prefix of the environment variables that will be present if
```
Comment (kumare3) [inline protos/flyteidl/core/security.proto:55]: do we want to change it to TOKENS?

**Review by kumare3 — COMMENTED** [review]
one question

File: protos/flyteidl/core/security.proto (hunk 6 above)
Comment (EngHabu) [reply protos/flyteidl/core/security.proto:57]: ummm change what to tokens?

File: protos/flyteidl/core/security.proto (hunk 5 above)
Comment (EngHabu) [reply protos/flyteidl/core/security.proto]: done

**Review by kumare3 — APPROVED** [review]
